lang/src/accounts/option.rs

```diff
@@ -0,0 +1,77 @@
+//! Option<T> type for optional accounts.
+//!
+//! # Example
+//! ```ignore
+//! #[derive(Accounts)]
+//! pub struct Example {
+//!     pub my_acc: Option<Account<'info, MyData>>
+//! }
+//! ```
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use solana_program::account_info::AccountInfo;
+use solana_program::instruction::AccountMeta;
+use solana_program::pubkey::Pubkey;
+
+use crate::{Accounts, AccountsClose, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+
+impl<'info, T: Accounts<'info>> Accounts<'info> for Option<T> {
+    fn try_accounts(
+        program_id: &Pubkey,
+        accounts: &mut &[AccountInfo<'info>],
+        ix_data: &[u8],
+        bumps: &mut BTreeMap<String, u8>,
+        reallocs: &mut BTreeSet<Pubkey>,
+    ) -> Result<Self> {
+        // There are three cases we have to handle. We don't care if
+        // accounts is empty, so if that's the case we return None. This
+        // allows adding optional accounts at the end of the Accounts
+        // struct without causing a breaking change. This is safe and will error
+        // out if a required account is then added after the optional account and
+        // the accounts aren't passed in.
+        if accounts.is_empty() {
+            return Ok(None);
+        }
+
+        // If there are enough accounts, it will check the program_id and return
+        // None if it matches, popping the first account off the accounts vec.
+        if accounts[0].key == program_id {
+            *accounts = &accounts[1..];
+            Ok(None)
+        } else {
+            // If the program_id doesn't equal the account key, we default to
+            // the try_accounts implementation for the inner type and then wrap that with
+            // Some. This should handle all possible valid cases.
+            T::try_accounts(program_id, accounts, ix_data, bumps, reallocs).map(Some)
+        }
+    }
+}
+
+impl<'info, T: ToAccountInfos<'info>> ToAccountInfos<'info> for Option<T> {
+    fn to_account_infos(&self) -> Vec<AccountInfo<'info>> {
+        self.as_ref()
+            .map_or_else(Vec::new, |account| account.to_account_infos())
```

Comment:
Thinking about this now, shouldn't this return the program id's account info? Maybe I'm just not familiar enough with the usage of `to_account_infos` and `to_account_metas`. But the `metas` one should probably also return the accountMeta of the program id?

Reply:
No, it looks like returning an empty vec from `to_account_infos` is probably fine. Since `to_account_infos` is mostly used when get the account infos from an account context to use in CPI.